[feature/main-travel > develop][#4] MainTravel 기본 기능 구현

ProtoSeo opened the pull request:
구현한 기능
- MainTravel Repository, Service, Controller 구현
  - MainTravel 단일 조회 기능
  - MainTravel 저장 기능
  - MainTravel 수정 기능
  - MainTravel 삭제 기능
 - 관련 단위 테스트 구현
TODO
- Member 작성자로 연결 하도록 구현
- Detail Travel 구현 후 `detailTravels` 에 추가하도록 구현
- Heart 구현 후 `hearts`에 이를 저장하도록 구현
- Scrap 구현 후 `scraps`에 이를 저장하도록 구현
- 이 후 테스트 수정

close #4

ProtoSeo committed:
[fix][#4] RuntimeException 상속 추가

ProtoSeo committed:
[feat][#4] MainTravelRepository 구현

ProtoSeo committed:
[feat][#4] MainTravel 관련 구현

- MainTravel 생성자 Builder 패턴 구현
- 수정을 위한 update 메서드 생성

ProtoSeo committed:
[feat][#4] MainTravel Service 관련 구현

- MainTravelService 기본적인 CRUD 관련 interface구현
- MainTravelService 구현체 구현

ProtoSeo committed:
[feat][#4] MainTravel Service계층 Dto 구현

- MainTravel 읽기 시 사용하는 MainTravelDto
- MainTravel 생성시 사용하는 MainTravelSaveDto
- MainTravel 수정시 사용하는 MainTravelUpdateDto

ProtoSeo committed:
[feat][#4] MainTravel Controller 계층 Dto 구현

- Update, Save 요청시 사용하는 MainTravelRequest 구현
- Response에 사용하는 MainTravelResponse 구현

ProtoSeo committed:
[feat][#4] MainTravelController 구현

- MainTravel 생성
- MainTravel 수정
- MainTravel 단일 조회
- MainTravel 삭제

ProtoSeo committed:
[feat][#4] MainTravel 예외관련 구현

ProtoSeo committed:
[refactor][#4] visible 오타 수정

ProtoSeo committed:
[fix][#4] MainTravelResponse, MainTravelDto isCompleted 추가

ProtoSeo committed:
[refactor][#4] MainTravelController 함수명 변경, 변수 명 변경

ProtoSeo committed:
[test][#4] MainTravel Test 관련 Fixture 생성

- MainTravel 도메인 Fixture 생성
- MainTravel Controller Dto Fixture 생성
- MainTravel Service Dto Fixture 생성

ProtoSeo committed:
[test][#4] MainTravelDto Test

- MainTravel 도메인에서 MainTravelDto로 변환 테스트

ProtoSeo committed:
[test][#4] MainTravelService 테스트 생성

- MainTravel 저장 테스트
- 저장된 MainTravel 불러오는 테스트
- MainTravel 삭제 테스트
- MainTravel 수정 테스트

ProtoSeo committed:
[test][#4] Controller dto 테스트

- Request 관련 테스트
- Response 관련 테스트

ProtoSeo committed:
[test][#4] MainTravel Controller 테스트

- MainTravel 저장 테스트
- MainTravel 단일 조회 테스트
- MainTravel 수정 테스트
- MainTravel 삭제 테스트

ProtoSeo committed:
Merge branch 'develop' into feature/main-travel

shin-mallang commented on src/main/java/com/footprint/maintravel/controller/MainTravelController.java:
```diff
@@ -0,0 +1,58 @@
+			.fromCurrentContextPath().path("/main-travels/")
+			.buildAndExpand(saveMainTravelId).toUri();
+
+		return ResponseEntity.created(location).body(saveMainTravelId);
```
Long을 반환하는 이유가 따로 있으신가요?

ProtoSeo replied:
이건 클라이언트에서 저장이나 업데이트 후에 어떤 화면을 보여줄 것인가에 따라 다른것같긴한데, 제 개인적인 생각으로는 저장 후에, 저장이 완료되었다는 화면이 나온 다음에 게시글 화면으로 이동할 때, 게시글의 id가 필요하지 않을까 해서 Long을 반환하도록 하였습니당!

shin-mallang commented on src/main/java/com/footprint/maintravel/controller/MainTravelController.java:
```diff
@@ -0,0 +1,58 @@
+	@DeleteMapping("/main-travels/{mainTravelId}")
+	public ResponseEntity<Void> deleteMainTravel(@PathVariable Long mainTravelId) {
+		mainTravelService.deleteMainTravel(mainTravelId);
+		return ResponseEntity.noContent().build();
```
Delete 요청이 성공한거라 noContent 대신 ok가 낫다고 생각하는데 어떻게 생각하시나요?

ProtoSeo replied:
<img width="792" alt="image" src="https://user-images.githubusercontent.com/55012742/169681712-7ed4eed2-07f8-4467-a821-09071c3d49ac.png">

[MDN 200 OK](https://developer.mozilla.org/ko/docs/Web/HTTP/Status/200) 보고 No Content를 사용했는데, 
<img width="804" alt="image" src="https://user-images.githubusercontent.com/55012742/169681761-478507e8-9914-44a4-b4a4-f87ea588dfae.png">

이걸 보고나서, 저희는 삭제하고 클라이언트의 페이지가 옮겨지니까 ok가 더 좋을 것같네요! 수정하도록 하겠습니당

shin-mallang commented on src/main/java/com/footprint/maintravel/controller/MainTravelController.java:
```diff
@@ -0,0 +1,58 @@
+public class MainTravelController {
+	private final MainTravelService mainTravelService;
+
+	@GetMapping("/main-travels/{mainTravelId}")
```
mainTravelId 대신 id로 적는 것은 어떻게 생각하시나요?
저는 MainTravelController이기 때문에 mainTravel이 굳이 붙지 않아도 명확하다 생각합니다

ProtoSeo replied:
좋은 의견인것 같습니당 수정하겠습니다!

shin-mallang commented on src/main/java/com/footprint/maintravel/controller/MainTravelController.java:
```diff
@@ -0,0 +1,58 @@
+		return ResponseEntity.ok(MainTravelResponse.from(mainTravelDto));
+	}
+
+	@PatchMapping("/main-travels/{mainTravelId}")
```
Put이 아닌 Patch를 사용하신 이유가 있으신가요?

ProtoSeo replied:
여기서는 Put이 맞을것같아요! Patch는 부분부분 변경할 때 사용하는 것이네용.. 제가 put이랑 patch랑 헷갈렸습니다..ㅜ 수정하도록 하겠습니당~

shin-mallang commented on src/main/java/com/footprint/maintravel/controller/dto/MainTravelRequest.java:
```diff
@@ -0,0 +1,34 @@
+import com.footprint.maintravel.service.dto.MainTravelSaveDto;
+import com.footprint.maintravel.service.dto.MainTravelUpdateDto;
+
+public record MainTravelRequest(
```
MainTravel을 생성할때 사용하는 Dto와 수정할때 사용하는 Dto를 각각 만들어서 관리하는건 어떻게 생각하시나요?

ProtoSeo replied:
Save할 때랑 Update할 때 body 내용이 똑같을것 같아서, request dto로만 으로도 충분할 것 같아요!

shin-mallang commented on src/main/java/com/footprint/maintravel/domain/MainTravel.java:
```diff
@@ -79,4 +78,21 @@ public class MainTravel {
+		this.likeNum = 0;
+	}
+
+	public void update(String title, LocalDate startDate, LocalDate endDate, Boolean isVisible, Boolean isCompleted) {
```
update 처리를 하는 방식에 대해 논의해봐야 할 것 같습니다

ProtoSeo replied:
넹! 일단 지금은 수정하지 않고 남겨두겠습니당. 만약 update에서, dto에 들어오는 값을 null 체크하고 하나씩 저장하게 된다면, save 단계에서도 null 체크를 해야할 것 같아요!

shin-mallang commented on src/main/java/com/footprint/maintravel/service/MainTravelService.java:
```diff
@@ -0,0 +1,16 @@
+
+public interface MainTravelService {
+
+	Long saveMainTravel(MainTravelSaveDto saveDto);
```
마찬가지로 Long 을 반환하는 이유가 있을까요?

ProtoSeo replied:
위에 참고해주세요!

shin-mallang commented on src/main/java/com/footprint/maintravel/service/MainTravelServiceImpl.java:
```diff
@@ -0,0 +1,59 @@
+
+	@Override
+	public Long saveMainTravel(MainTravelSaveDto saveDto) {
+		MainTravel save = mainTravelRepository.save(MainTravel.builder()
```
해당 과정을 MainTravelSaveDto에 추가하여 saveDto.toEntity() 형식으로 바꾸는 것은 어떻게 생각하시나요?

ProtoSeo replied:
괜찮을것같아요~ 수정하도록 하겠습니당~

shin-mallang commented on src/main/java/com/footprint/maintravel/service/MainTravelServiceImpl.java:
```diff
@@ -0,0 +1,59 @@
+
+	@Override
+	public void deleteMainTravel(Long mainTravelId) {
+		MainTravel mainTravel = mainTravelRepository.findById(mainTravelId).orElseThrow();
```
여기서 예외처리 따로 안하신 이유가 있으신가요?

ProtoSeo replied:
빼놓고 구현했네용,, 추가했습니당!

shin-mallang commented on src/main/java/com/footprint/maintravel/service/dto/MainTravelDto.java:
```diff
@@ -0,0 +1,25 @@
+
+import com.footprint.maintravel.domain.MainTravel;
+
+public record MainTravelDto(
```
createdDate도 포함되어야 하지 않을까요?

ProtoSeo replied:
ERD랑 도메인에도 없어서, 일단 도메인이랑 ERD에 추가하고 추가하도록 하겠습니당~

shin-mallang reviewed (changes requested):
확인해주시면 감사하겠습니다 :D

ProtoSeo committed:
[refactor][#4] delete api에서 Httpstatus no content에서 ok로 수정

ProtoSeo committed:
[refactor][#4] mainTravelId -> id로 수정

ProtoSeo committed:
[refactor][#4] MainTravel 수정 시 HttpMethod 변경

- 수정 시, 리소스(MainTravel)를 덮는 방식으로 수정하므로, PutMapping이 더 적합하다.

ProtoSeo committed:
[refactor][#4] MainTravel 삭제 시 예외처리 추가

- MainTravel ServiceImpl 에서 MainTravel 삭제 시 예외처리 추가
- MainTravel Exception static import 로 간결하게 추가

ProtoSeo committed:
[refactor][#4] saveDto에 Entity변환 메서드 추가

ProtoSeo committed:
[test][#4] saveDto에서 Entity 변환 테스트 구현

ProtoSeo committed:
[refactor][#4] BaseTimeEntity 상속 추가

ProtoSeo committed:
[refactor][#4] Dto CreatedAt 추가

- MainTravelResponse Dto CreatedAt 추가
- MainTraveDto CreatedAt 추가

ProtoSeo committed:
[test][#4] CreatedAt 관련 수정

ProtoSeo committed:
[test][#4] Dto createdAt 관련 테스트 추가

ProtoSeo committed:
Merge branch 'feature/main-travel' of https://github.com/AppDev-FootPrint/FootPrint-backend into feature/main-travel
